qiita_pet/handlers/api_proxy/sample_template.py

```diff
@@ -0,0 +1,269 @@
+    cols = list(df.columns)
+    for column in cols:
+        counts = df[column].value_counts()
+        out['summary'][str(column)] = [(str(key), counts[key])
+                                       for key in natsorted(counts.index)]
+
+    return out
+
+
+@execute_as_transaction
+def sample_template_post_req(study_id, user_id, data_type,
+                             sample_template):
+    """Creates the sample template from the given file
+
+    Parameters
+    ----------
+    study_id : int
+        The current study object id
+    user_id : str
+        The current user object id
+    data_type : str
+        Data type for the sample template
+    sample_template : str
+        filepath to use for creation
+
+    Returns
+    -------
+    dict
+        results dictonary in the format
+        {'status': status,
+         'message': msg,
+         'file': sample_template}
+
+    status can be success, warning, or error depending on result
+    message has the warnings or errors
+    file has the file name
+    """
+    access_error = check_access(int(study_id), user_id)
+    if access_error:
+        return access_error
+    # Get the uploads folder
+    _, base_fp = get_mountpoint("uploads")[0]
+    # Get the path of the sample template in the uploads folder
+    fp_rsp = join(base_fp, str(study_id), sample_template)
+
+    if not exists(fp_rsp):
+        # The file does not exist, fail nicely
+        return {'status': 'error',
+                'message': 'filepath does not exist',
+                'file': sample_template}
+
+    # Define here the message and message level in case of success
+    msg = ''
+    status = 'success'
+    is_mapping_file = looks_like_qiime_mapping_file(fp_rsp)
+    if is_mapping_file and not data_type:
+        return {'status': 'error',
+                'message': 'Please, choose a data type if uploading a '
+                           'QIIME mapping file',
+                'file': sample_template}
+
+    study = Study(int(study_id))
+    try:
+        with warnings.catch_warnings(record=True) as warns:
+            if is_mapping_file:
+                create_templates_from_qiime_mapping_file(fp_rsp, study,
+                                                         int(data_type))
+            else:
+                SampleTemplate.create(load_template_to_dataframe(fp_rsp),
+                                      study)
+            remove(fp_rsp)
+
+            # join all the warning messages into one. Note that this
+            # info will be ignored if an exception is raised
+            if warns:
+                msg = '; '.join([convert_text_html(str(w.message))
+                                 for w in warns])
+                status = 'warning'
+
+    except Exception as e:
+        # Some error occurred while processing the sample template
+        # Show the error to the user so they can fix the template
+        status = 'error'
+        msg = str(e)
+    return {'status': status,
```

Comment:
It's returning the filepath, not the file itself. This allows better messaging to users and anything else that acts on the system, since you know not only if it was successful, but what the function is acting on.